huaweicloud/services/iam/resource_huaweicloud_identity_service_agency.go

```diff
@@ -0,0 +1,539 @@
+		if err != nil {
+			return diag.Errorf("error updating IAM service agency: %s", err)
+		}
+	}
+
+	// update attached policy
+	if d.HasChange("policy_names") {
+		oRaw, nRaw := d.GetChange("policy_names")
+		oMap := oRaw.(*schema.Set)
+		nMap := nRaw.(*schema.Set)
+
+		// list policy IDs which will be detached
+		detachPolicyIDs, err := getPolicyIDsByNames(client, oMap.Difference(nMap).List())
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		// detach policy by ID
+		for _, detachPolicyID := range detachPolicyIDs {
+			if err := detachPolicyByID(client, d.Id(), detachPolicyID); err != nil {
+				return diag.FromErr(err)
+			}
+		}
+
+		// list policy IDs which will be attached
+		attachPolicyIDs, err := getPolicyIDsByNames(client, nMap.Difference(oMap).List())
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		// attach policy by ID
+		for _, attachPolicyID := range attachPolicyIDs {
+			if err := attachPolicyByID(client, d.Id(), attachPolicyID); err != nil {
+				return diag.FromErr(err)
+			}
+		}
+	}
+
+	// update tags
+	if d.HasChange("tags") {
+		oRaw, nRaw := d.GetChange("tags")
+		oMap := oRaw.(map[string]interface{})
+		nMap := nRaw.(map[string]interface{})
+
+		// remove old tags
+		if len(oMap) > 0 {
+			if err = deleteTags(client, oMap, "agency", d.Id()); err != nil {
+				return diag.FromErr(err)
+			}
+		}
+
+		// set new tags
+		if len(nMap) > 0 {
+			if err := createTags(client, nMap, "agency", d.Id()); err != nil {
+				return diag.FromErr(err)
+			}
+		}
+	}
+
+	return resourceIAMServiceAgencyRead(ctx, d, meta)
+}
+
+func getPolicyIDsByNames(client *golangsdk.ServiceClient, policyNames []interface{}) ([]string, error) {
+	listPolicyHttpUrl := "v5/policies"
+	listPolicyPath := client.Endpoint + listPolicyHttpUrl
+	listPolicyPath += fmt.Sprintf("?limit=%v", pageLimit)
+	listPolicyOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+	}
+
+	path := listPolicyPath
+	rst := make([]string, 0)
+	for {
+		listPolicyResp, err := client.Request("GET", path, &listPolicyOpt)
+		if err != nil {
+			return nil, fmt.Errorf("error getting IAM policies: %s", err)
+		}
+		listPolicyRespBody, err := utils.FlattenResponse(listPolicyResp)
+		if err != nil {
+			return nil, err
+		}
+
+		policies := utils.PathSearch("policies", listPolicyRespBody, make([]interface{}, 0))
+		for _, policy := range policies.([]interface{}) {
+			policyName := utils.PathSearch("policy_name", policy, "")
+			if policyName == "" {
+				return nil, fmt.Errorf("error getting policy name")
+			}
+			policyID := utils.PathSearch("policy_id", policy, "")
+			if policyID == "" {
+				return nil, fmt.Errorf("error getting policy ID for name(%s)", policyName)
+			}
+
+			// use policy name to filter result
+			if utils.StrSliceContains(utils.ExpandToStringList(policyNames), policyName.(string)) {
+				rst = append(rst, policyID.(string))
+			}
+		}
+
+		marker := utils.PathSearch("page_info.next_marker", listPolicyRespBody, "")
```

Comment:
we should break when all names were found.
```suggestion
		if len(rst) == len(policyNames) {
				break
		}

		marker := utils.PathSearch("page_info.next_marker", listPolicyRespBody, "")
```